check_pgactivity

```diff
@@ -5600,6 +5604,167 @@ sub check_oldest_idlexact {
+            ORDER BY age(transaction) DESC NULLS LAST
+          ) rownum, age(transaction) AS age, transaction
+          FROM (SELECT transaction FROM pg_prepared_xact
+            UNION ALL SELECT NULL
+          ) sql
+        )
+        SELECT kind, age, xmin FROM ordered
+        WHERE rownum = 1
+        },
+        # backend_xmin and backend_xid added to pg_stat_activity,
+        # backend_xmin added to pg_stat_replication,
+        # replication slots introduced
+        $PG_VERSION_94 => q{
+        WITH raw AS (
+          -- regular backends
+          SELECT
+          CASE WHEN xact_start = query_start
+            THEN 'query'
+            ELSE
+              CASE WHEN state = 'idle in transaction'
+                THEN 'idle_xact'
+                ELSE 'active_xact'
+              END
+          END AS kind,
+          coalesce(backend_xmin, backend_xid) AS xmin
+          FROM pg_stat_activity
+          UNION ALL (
+            -- 2PC
+            SELECT '2pc' AS kind,
+            transaction AS xmin
+            FROM  pg_prepared_xacts
+            WHERE database = current_database()
+          ) UNION ALL (
+            -- replication slots
+            SELECT 'repslot' AS kind, xmin AS xmin
+            FROM  pg_replication_slots
+            WHERE coalesce(database, current_database()) = current_database()
+          ) UNION ALL (
+            -- walsenders
+            SELECT 'walsender' AS kind, backend_xmin AS xmin
+            FROM pg_stat_replication
+          )
+        ),
+        ordered AS (
+          SELECT f.kind,
+          -- xid type doesn't have range operators as the value will wraparound.
+          -- Instead, rely on age() function and row_number() window function
+          -- to get the oldest xid found.
+          row_number() OVER (
+            PARTITION BY f.kind
+            ORDER BY age(xmin) DESC NULLS LAST
+          ) rownum, age(xmin) AS age, xmin
+          FROM raw
+          RIGHT JOIN (
+            SELECT           'query'
+            UNION ALL SELECT 'idle_xact'
+            UNION ALL SELECT 'active_xact'
+            UNION ALL SELECT '2pc'
+            UNION ALL SELECT 'repslot'
+            UNION ALL SELECT 'walsender'
+          ) f(kind) ON raw.kind = f.kind
+        )
+        SELECT kind, age, xmin FROM ordered
+        WHERE rownum = 1
+        }
+    );
+
+    # warning and critical must be raw.
+    pod2usage(
+        -message => "FATAL: critical and warning thresholds only accept raw number of transactions.",
+        -exitval => 127
+    ) unless $args{'warning'}  =~ m/^([0-9.]+)$/
+        and  $args{'critical'} =~ m/^([0-9.]+)$/;
+
+    $c_limit = $args{'critical'};
+    $w_limit = $args{'warning'};
+
+    @hosts = @{ parse_hosts %args };
+
+    pod2usage(
+        -message => 'FATAL: you must give only one host with service "oldest_xmin".',
+        -exitval => 127
+    ) if @hosts != 1;
+
+    is_compat $hosts[0], 'oldest_xmin', $PG_VERSION_84 or exit 1;
```

Comment:
I wonder what is the point to support 9.3 and before as we only check for prepared xact. There's already the `oldest_2pc` check that raise an alert for stalled prepared xact.

Reply:
Because a 1m old 2pc holding 10 transactions is not the same as a 1min old 2pc holding 1M transactions, and if you have a non uniform xid generation it's likely to have random bloat issue that oldest_2pc won't be able to help without a lot of noise?

Reply:
Well, ok.

FYI, I can test all versions up to 8.1 where pg_prepared_xacts has been introduced. So you can go the full way down this abyssal version.

Reply:
Thanks!